ale/util.py

```diff
@@ -506,3 +501,303 @@ def query_kernel_pool(matchstr="*"):
+
+            next_meta = [(i,m) for i,m in enumerate(metadata[imeta:]) if m[1] == "object_end"][0]
+            next_group_start = next_meta[1][0]
+
+            group_lines = lines[element_start_line+1:next_group_start]
+            items.append((group_name, JBFPvlParser(group_lines)))
+            imeta += next_meta[0]
+        elif element_type == "object_end" or element_type == "group_end":
+            imeta+=1
+
+    return pvl.PVLModule(items)
+
+
+def search_isis_db(dbobj, labelobj, isis3_data="/usgs/cpkgs/isis3/data/"):
+    """
+    Given an PVL obj of a KernelDB file and an Isis Label for a cube, find the best kernel
+    to attach to the cube.
+
+    The Logic here is a bit gross, but it matches ISIS's implementation very closely.
+
+
+    Parameters
+    ----------
+    dbobj : PVLModule
+            ISIS3 KernelDB file as a loaded PVLModule
+
+    labelobj : PVLModule
+               Cube label as loaded PVLModule
+
+    isis3_data : str
+                 path to $ISIS3DATA
+
+    Returns
+    -------
+    : dict
+      dictionary containing kernel list and optionally the kernel type if relevant.
+
+    """
+    if not dbobj:
+        return
+
+    utc_start_time = labelobj["IsisCube"]["Instrument"]["StartTime"]
+    utc_stop_time = labelobj["IsisCube"]["Instrument"]["StopTime"]
+
+    run_time = None
+    dependencies = None
+    kernels = []
+    typ = None
+    types = []
+
+    # flag is set when a kernel is found matching the start time but not stop time,
+    # this may
+    partial_match = False
+    full_match = False
+
+    for selection in dbobj.getlist("Selection"):
+        files = selection.getlist("File")
+
+        # selection criteria
+        matches = selection.getlist("Match")
+        times = selection.getlist("Time")
+
+        if not files:
+            raise Exception(f"No File found in {pvlobj}")
+
+        files = [path.join(*file) if isinstance(file, list) else file  for file in files]
+
+        for i,time in enumerate(times):
+            isis_time_format = '%Y %b %d %H:%M:%S.%f TDB'
+            time = (datetime.strptime(time[0], isis_time_format),
+                               datetime.strptime(time[1], isis_time_format))
+            start_time_in_range = utc_start_time >= time[0] and utc_start_time <= time[1]
+            stop_time_in_range = utc_stop_time >= time[0] and utc_stop_time <= time[1]
+            times[i] = stop_time_in_range, stop_time_in_range
+
+        for i,match in enumerate(matches):
+            matches[i] = labelobj["IsisCube"][match[0]][match[1]].lower() == match[2].lower()
+
+        if any(matches if matches else [True]):
+            for i,f in enumerate(files):
+                if isinstance(f, tuple):
+                    f = os.path.join(*f)
+
+                full_path = os.path.join(isis3_data, f).replace("$", "")
+                if '?' in full_path:
+                    full_path = sorted(glob(full_path))[-1]
+                files[i] = full_path
```

Comment:
There are some files that use more complex Isis::FileName matching. For example, here's what the Cassini PCK selection looks like:

```
  Group = Selection
    Time = ("1999 AUG 19 02:00:00.000 TDB", "2001 MAR 24 00:00:00.000 TDB")
    File = (base, kernels/pck/pck?????.tpc)
    File = "$cassini/kernels/pck/cpck{ddMMMyyyy}_jupiter.tpc"
  End_Group
```